File: RegexFileSorter/Sorter.vb
Imports System.IO
Imports System.Text.RegularExpressions
Imports RegexFileSorter.Configuration

Public Class Sorter

	Public Shared Function ScanFiles() As ILookup(Of String, String)
		Dim Result As New List(Of (Path As String, Match As String))
		Dim S = Directory.GetFiles(Current.SFolder)
		Dim R = New Regex(Current.Regex)
		For Each File In S
			Dim M = R.Match(Path.GetFileName(File))
			If M.Success Then
				'Result.Add(New MatchedFile With {.Path = File, .Match = M.Groups("S").Value})
				Result.Add((File, M.Groups("S").Value))
			End If
		Next
		Return Result.ToLookup(Function(x) x.Match, Function(x) x.Path)
	End Function

	Public Shared Function PrepareFiles(G As IGrouping(Of String, String)) As SortedFolder
		Dim SF = New SortedFolder(G.Key)
		SF.Path = Path.Combine(Current.OutFolder, SF.Name)
		If Directory.Exists(SF.Path) Then
			SF.IsValid = True
			SF.Status = "Found"
		ElseIf Current.SearchFolder Then
			For Each Folder In Directory.GetDirectories(Current.OutFolder)
				Dim FolderName = Path.GetFileName(Folder)
				If FolderName.StartsWith(SF.Name, StringComparison.CurrentCultureIgnoreCase) Then
					SF.Path = Folder
					SF.IsValid = True
					SF.Status = $"Found {FolderName}"
					Exit For
				End If
			Next
		End If
		If Not SF.IsValid Then
			If Current.CreateNew Then
				SF.IsValid = True
				SF.Status = $"Create folder {SF.Name}"
			Else
				SF.Status = "Not found"
			End If
		End If

		For Each File In G
			Dim FileName = Path.GetFileName(File)
			Dim OutFile = Path.Combine(SF.Path, FileName)
			SF.Files.Add(New SortedFolder.SortedFile(FileName, File, OutFile))
		Next
		Return SF
	End Function

	Public Shared Sub MoveFiles(SFList As IEnumerable(Of SortedFolder))
		For Each SF In SFList
			If Not SF.IsValid Then Continue For
			For Each F In SF.Files
				File.Move(F.InPath, F.OutPath)
			Next
		Next
	End Sub

End Class

File: RegexFileSorterSharp/Sorter.cs
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;
using static RegexFileSorter.Configuration;

namespace RegexFileSorter
{
    internal static class Sorter
    {
        public static void MoveFiles(IEnumerable<SortedFolder> folders)
        {
            foreach (var SF in folders)
            {
                if (!SF.IsValid) { continue; }
                foreach (var F in SF.Files)
                {
                    if (File.Exists(F.OutPath)) { File.Delete(F.OutPath); }
                    File.Move(F.InPath, F.OutPath);
                }
            }
        }

        public static SortedFolder PrepareFiles(IGrouping<string, string> group)
        {
            var SF = new SortedFolder(group.Key);
            SF.Path = Path.Combine(Current.OutFolder, SF.Name);
            if (Directory.Exists(SF.Path))
            {
                SF.IsValid = true;
                SF.Status = "Found";
            }
            else if (Current.SearchFolder)
            {
                foreach (var Folder in Directory.GetDirectories(Current.OutFolder))
                {
                    var FolderName = Path.GetFileName(Folder);
                    if (FolderName.ToLower().Contains(SF.Name.ToLower()))
                    {
                        SF.Path = Folder;
                        SF.IsValid = true;
                        SF.Status = $"Found \"{FolderName}\"";
                    }
                }
            }

            if (!SF.IsValid)
            {
                if (Current.CreateNew)
                {
                    SF.IsValid = true;
                    SF.Status = $"Create folder \"{SF.Name}\"";
                }
                else { SF.Status = "Not Found"; }
            }

            foreach (var File in group)
            {
                var FileName = Path.GetFileName(File);
                var OutFile = Path.Combine(SF.Path, FileName);
                SF.Files.Add(new SortedFolder.SortedFile(FileName, File, OutFile));
            }

            return SF;
        }

        public static ILookup<string, string> ScanFiles()
        {
            var Result = new List<(string Path, string Match)>();
            var S = Directory.GetFiles(Current.SFolder);
            var R = new Regex(Current.Regex);
            foreach (var File in S)
            {
                var M = R.Match(Path.GetFileName(File));
                if (M.Success) { Result.Add((File, M.Groups["S"].Value)); }
            }
            return Result.ToLookup(X => X.Match, X => X.Path);
        }
    }
}